### PR TITLE
Simplify api

### PR DESCRIPTION
**Pull request by james-lawrence**
to get the conversation started. this PR is mostly a proof of concept/starting point for cleaning up the client API. doesn't implement the underlying improvements that could be made.

### CODE REVIEW

File: cmd/torrent-pick/main.go [hunk 1]
```diff
@@ -136,24 +136,24 @@ func main() {
...
-					log.Fatal(err)
-				}
-				t, err := client.AddTorrent(metaInfo)
-				if err != nil {
-					log.Fatal(err)
-				}
-				return t
 			}
+
+			metaInfo, err := metainfo.LoadFromFile(arg)
```
Comment (james-lawrence) [inline cmd/torrent-pick/main.go]: could be further simplified:
```
t, _, err := client.MaybeStart(torrent.NewFromMetainfoFile(arg))
```

File: spec.go [hunk 2]
```diff
@@ -1,13 +1,54 @@
...
+
+// OptionChunk sets the size of the chunks to use for outbound requests
+func OptionChunk(s int) Option {
+	return func(t *Torrent) {
+		t.ChunkSize = s
+	}
+}
+
+// OptionStorage set the storage implementation for the torrent.
+func OptionStorage(s storage.ClientImpl) Option {
```
Comment (james-lawrence) [inline spec.go]: could break the package dependency here by defining its own private interface for storage to satisfy.

File: client.go [hunk 3]
```diff
@@ -69,28 +71,155 @@ type Client struct {
...
+		return dl, false, failed
+	}
+	return cl.start(t)
+}
+
+// Download the specified torrent.
+// Download differs from Start in that it blocks until the torrent is completely downloaded.
+// the provided ctx is used to timeout the download, but it currently only works with regards
+// to fetching the torrent info.
+func (cl *Client) Download(ctx context.Context, t Torrent) (dl Download, added bool, err error) {
```
Comment (james-lawrence) [inline client.go]: totally hacky download method. details here can be cleaned up pretty sure the copy is unnecessary.

File: client.go [hunk 4]
```diff
@@ -69,28 +71,155 @@ type Client struct {
...
+// Stop the specified torrent, this halts all network activity around the torrent
+// for this client.
+func (cl *Client) Stop(t Torrent) (err error) {
+	cl.lock()
+	defer cl.unlock()
+	return cl.dropTorrent(t.InfoHash)
+}
+
+// SetMaxEstablishedConns for the torrent.
+func (cl *Client) SetMaxEstablishedConns(t Torrent, max int) (oldMax int) {
```
Comment (james-lawrence) [inline client.go]: this is a minor artifact where I was trying to reduce the number of methods publicly exposed on the original Torrent struct.

leaning towards changing this method to something along the lines of:
```golang
func (cl *Client) Tune(t Torrent, options ...Tuning) error {
  // applying tuning configuration to the specified torrent.
}
```

File: client.go [hunk 5]
```diff
@@ -69,28 +71,155 @@ type Client struct {
...
+// creation methods: New, NewFromFile, etc. it is all respects identical to the Start
+// method.
+func (cl *Client) MaybeStart(t Torrent, failed error) (dl Download, added bool, err error) {
+	if failed != nil {
+		return dl, false, failed
+	}
+	return cl.start(t)
+}
+
+// Download the specified torrent.
```
Comment (james-lawrence) [inline client.go]: I believe a sister method to Download should be added. similar in nature it blocks until the torrents Info is retrieved,
or the context expires.
```golang
func (cl *Client) Info(ctx context.Context, t Torrent) error {
  // ...
}
```

File: client.go [hunk 6]
```diff
@@ -485,9 +614,9 @@ type dialResult struct {
 
 func countDialResult(err error) {
 	if err == nil {
-		torrent.Add("successful dials", 1)
+		torrentVar.Add("successful dials", 1)
```
Comment (james-lawrence) [inline client.go]: can likely clean most of this noise up by changing `type torrent` to `type download`

**Comment by anacrolix** [comment]
Thanks for the PR. Consider making some of these changes in separate PRs, and discussing them on gitter or in an issue first.

**Comment by james-lawrence** [comment]
@anacrolix - issue #243, I've broached this topic with you via an issue and in glitter. getting you to respond in a collaborative manner has been rather difficult. =)

i made the PR to focus the discussion on actual code. do you actually have any suggestions to make? how would you like it broken up in separate PRs? what aprts did you like/dislike? any feedback at all is useful.

**Comment by anacrolix** [comment]
@james-lawrence Let's address this in #243.

**Pull request closed without merging**